qt: Get rid of cursor in out-of-focus labels

hebasto opened the pull request:
After clicking on `QLabel` with selectable text the cursor remains forever:

![47532924-65e7b200-d8ba-11e8-9254-7bde658961cb](https://user-images.githubusercontent.com/32963518/84038485-ad945200-a9a8-11ea-89e3-c7c17d02a611.png)

This PR fixes this visual bug.

Earlier attempts to fix this issue:
- #14577
- #14810 (combined with other UX feature)

hebasto committed:
qt: Get rid of cursor in out-of-focus labels

This change is a temporary fix of QTBUG-59514.

laanwj commented:
Concept ACK, multiple cursors visible is wtf-level  confusing.

Should this be reported as an upstream issue too?

promag reviewed (commented):
Concept ACK. Tested this approach and the one I'm suggesting. Does it make sense to have a dedicated label subclass?

luke-jr commented:
This seems ugly. Ideally, if the focus returns to the label, the cursor position shouldn't be lost.

hebasto commented:
@laanwj 
> Should this be reported as an upstream issue too?

https://bugreports.qt.io/browse/QTBUG-59514

hebasto commented:
Updated 4b7c95324f40738185f52da161d144f6e9f0c484 -> 5dbe2d60a83c06d550f104d79e1b9c82e49c7008 ([pr19210.01](https://github.com/hebasto/bitcoin/commits/pr19210.01) -> [pr19210.02](https://github.com/hebasto/bitcoin/commits/pr19210.02), [diff](https://github.com/hebasto/bitcoin/compare/pr19210.01..pr19210.02)):

- addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/19210#discussion_r438432468)
> This also works

- fixed context menu event handling

hebasto commented on src/qt/guiutil.cpp:
```diff
@@ -450,6 +450,25 @@ bool ToolTipToRichTextFilter::eventFilter(QObject *obj, QEvent *evt)
+    if (event->type() == QEvent::FocusOut) {
+        QLabel* label = qobject_cast<QLabel*>(watched);
+        if (label) {
+            const QString content = label->text();
```
It looks more effective. Thanks!

[Updated](https://github.com/bitcoin/bitcoin/pull/19210#issuecomment-653585810).

hebasto commented:
@promag 
> Does it make sense to have a dedicated label subclass?

I think, it does not required.

@luke-jr 
> This seems ugly. Ideally, if the focus returns to the label, the cursor position shouldn't be lost.

Unfortunately, the `QLabel` class has no property that holds the cursor position.

promag commented on src/qt/bitcoin.cpp:
```diff
@@ -556,6 +556,9 @@ int GuiMain(int argc, char* argv[])
     // Install global event filter that makes sure that long tooltips can be word-wrapped
     app.installEventFilter(new GUIUtil::ToolTipToRichTextFilter(TOOLTIP_WRAP_THRESHOLD, &app));
+    // Install global event filter that makes sure that out-of-focus labels do not contain text cursor.
+    // This is a temporary fix of QTBUG-59514.
```
nit, move/copy this comment to class doc.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19210#issuecomment-653598259).

promag reviewed (commented):
Code review ACK 5dbe2d60a83c06d550f104d79e1b9c82e49c7008.

hebasto commented:
Updated 5dbe2d60a83c06d550f104d79e1b9c82e49c7008 -> bd315eb5e27d49d47759ae9417328427426cb269 ([pr19210.02](https://github.com/hebasto/bitcoin/commits/pr19210.02) -> [pr19210.03](https://github.com/hebasto/bitcoin/commits/pr19210.03), [diff](https://github.com/hebasto/bitcoin/compare/pr19210.02..pr19210.03)):

- addressed @promag's [comment](https://github.com/bitcoin/bitcoin/pull/19210#discussion_r449632417)
> nit, move/copy this comment to class doc

promag commented:
Code review ACK bd315eb5e27d49d47759ae9417328427426cb269.

fanquake commented:
@laanwj can you take another look here? If you're happy with this change as-is, we can probably merge is here rather than punting to the GUI repo. We've now also got an upstream issue to refer too.

laanwj commented:
Tested ACK bd315eb5e27d49d47759ae9417328427426cb269
